### ISSUE TITLE
Serialize a dictionary.

### ISSUE THREAD

**Issue by isaac-weisberg** [dhlab-basel/json2typescript#68]
I would like to serialize a dictionary of type { [id: string]: Foo }
What should I do?...

**Comment by isaac-weisberg** [comment]
Actually, I will attempt to write a custom initializer

**Comment by isaac-weisberg** [comment]
Alright, so I did that

```typescript
@JsonObject class Foo {

}

@JsonObject class Dictionary<Type> {
    [id: string]: Foo
}

@JsonConverter class DictionaryConverter implements JsonCustomConvert<Dictionary<Foo>> {
    serialize(dictionary: Dictionary<Foo>): any {
        let out = {}
        let convert = new JsonConvert()
        for (let key in dictionary) {
            let obj = dictionary[key]
            out[key] = convert.serialize(obj)
        }
        return out
    }

    deserialize(data: any): Dictionary<Foo> {
        let convert = new JsonConvert()
        let dict = new Dictionary<Foo>()
        for (let key in data) {
            let value: Foo = convert.deserialize(data[key], Foo)
            dict[key] = value
        }
        return dict
    }
}
```

But I would like to make a generic version of the `class DictionaryConverter`, but unfortunately I am not aware of any way to guarantee in a generic function that generic type parameter is a class constructor, and thus `let value: Type = convert.deserialize(data[key], Type)` where `Type` is generic type doesn't compile.

**Comment by andreas-aeschlimann** [comment]
The main issue about generics is that they are removed when you compile to JavaScript. As `json2typescript` runs on runtime, there is no direct way to handle it.

I think you are on the right track. You will find a good workaround that suits you well.

By the way, `json2typescript` can already map a dictionary to an array out of the box. But if I remember right, you will lose the keys that way.

**Comment by isaac-weisberg** [comment]
@andreas-aeschlimann yeah, your statement is as close as it gets, and it really looks like no generic serialization solution for simple key-typed value storage. So in the project I am dealing with, I am going to go for non generic `@JsonConverter`s for each kind of dictionary I need.

**Comment by isaac-weisberg** [comment]
So I've been fiddling around...
```typescript
import { JsonObject, JsonConvert, JsonProperty, JsonConverter, JsonCustomConvert } from 'json2typescript'

@JsonObject class Dictionary<Type> {
    [id: string]: Type
}

@JsonConverter
class DictionaryConverter<Type> implements JsonCustomConvert<Dictionary<Type>> {
    static iden: string = ""
    static gets = 0
    static constructors: {[id: string]: any[]} = {}

    static begin(iden: string) {
        this.iden = iden
        this.gets = 0
    }

    static getConstructor<GenericType>(type: new () => GenericType, iden: string): new () => DictionaryConverter<GenericType> {
        let array: any[] = this.constructors[iden]
        if (!array) {
            array = []
            this.constructors[iden] = array
        }
        array.push(type)
        return DictionaryConverter
    }

    static getNextType(): any {
        let constr = this.constructors[this.iden][this.gets]
        this.gets++
        return constr
    }

    serialize(data: Dictionary<Type>): any {
        let convert = new JsonConvert()
        let obj: any = {}
        for (let key in data.storage) {
            let object = data[key]
            if (object) {
                obj[key] = convert.serialize(object)
            }
        }
        return obj
    }
    deserialize(data: any): Dictionary<Type> {
        let type = DictionaryConverter.getNextType()
        let dict = new Dictionary<Type>()
        let convert = new JsonConvert()
        for (let key in data) {
            dict[key] = convert.deserialize(data[key], type)
        }
        return dict
    }
}

@JsonObject class Foo {
    constructor(msg: string = "") {
        this.msg = msg
    }
    @JsonProperty("message", String)
    msg: string = ""
}

@JsonObject class Container {
    @JsonProperty("dict", DictionaryConverter.getConstructor(Foo, "Contain"))
    dict: Dictionary<Foo> = new Dictionary()
}

let convert = new JsonConvert()

let contain = new Container()
contain.dict["thingie"] = new Foo("Thingieee")
contain.dict["dingilingie"] = new Foo("dingilingie")

let functor = (container: Container) => {
    let raw = convert.serialize(contain)
    let jsonString = JSON.stringify(raw)
    console.log("The JSON I got")
    console.log(jsonString)
    let obj = JSON.parse(jsonString)
    DictionaryConverter.begin("Contain")
    let newContainer = convert.deserialize(obj, Container)
    return newContainer
}

var swapContain: Container = contain
for (let i = 0; i < 10; i++) {
    swapContain = functor(swapContain)
    console.log("Contain object")
    console.log(swapContain)
}
```

**Comment by isaac-weisberg** [comment]
https://github.com/isaac-weisberg/generic-dictionary
This is the result, my solution.

**Comment by isaac-weisberg** [comment]
doesn't support number.

**Comment by isaac-weisberg** [comment]
A couple of weeks later, heavily digging into the semantics of the project I required the aforementioned feature for, and the domain where the project is to be applied, I am summarizing and closing this issue because of this:

TL;DR: if you are going to write some serializable data management and plan to utilize context-switched, stateful data managers, while using javascript runtime to power the system with it's typing capabilities, just don't fucking bother with dictionaries, use an array of objects, each of which contains a key and filter/forEach as appropriate.

Actually, I don't even know why did I title the previous sentence as TL;DR, if that's basically what it is, it just doesn't support these guarantees you get in strongly typed languages, it's just plain objects with keys that you can iterate. Caveats and rakes to surpass limitations of js runtime. I am dissatisfied with my experience of trying to solve the issue.
